src/exclusion/object_filter.hpp

```diff
@@ -22,35 +24,131 @@
 namespace ddwaf::exclusion {
 
 class path_trie {
-public:
-    path_trie() = default;
-
-    [[nodiscard]] path_trie find(std::string_view key) const;
-
-    template <typename T> [[nodiscard]] path_trie find(const std::vector<T> &path) const;
-
-    template <typename T> void insert(const std::vector<T> &path);
-
-    [[nodiscard]] bool is_terminal() const { return root && root->terminal; }
-    [[nodiscard]] bool is_valid() const { return root && !root->values.empty(); }
-
-protected:
+private:
     struct trie_node {
-        // Heterogenous lookups required...
-        std::unordered_map<std::string_view, std::shared_ptr<trie_node>> values;
-        // TODO this shouldn't be necessary
-        bool terminal{false};
+        std::string path_component;
+        std::vector<trie_node> children{};
```

Comment:
You mean with `struct trie_node { vector<pair<string, trie_node>> children; }`? I'm only seeing drawbacks: the pair impairs the legibility (the first template argument of the pair is not given a name), the iterator gets bigger because it needs to store some string reference for the path_component, and I probably I still need to provide a custom comparator.

Reply:
> As mentioned, I think the reduced cost of indirection will be dwarfed by the search cost.

I'll try to benchmark it.